torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -373,6 +377,36 @@ void fixDefaultLstmCellState(Block *b) {
   }
 }
 
+static bool isSafeToSpeculate(Node* n) {
+  return n->kind() == onnx::Transpose;
+}
+
+static void speculateOps(Block* block) {
+  for(auto it = block->nodes().begin(), end = block->nodes().end();
+      it != end;) {
+    Node * n = *it;
+    ++it; //note: increment first so that it is safe to move the node if needed
+
+    for(auto b : n->blocks()) {
+      speculateOps(b);
+    }
+    if(!isSafeToSpeculate(n))
+      continue;
+    // XXX - only works for nodes with a single input
+    // move node n outside of the control flow it is nested in
+    auto node_input = n->input()->node();
+    if(node_input->owningBlock() == n->owningBlock())
+      continue;
+    // find the control flow node in the same block as node_input that contains
+    // Node n
+    auto flow_node = n->owningBlock()->owningNode();
+    while(flow_node->owningBlock() != node_input->owningBlock())
+      flow_node = flow_node->owningBlock()->owningNode();
+    // put the node right before this flow node
+    n->moveBefore(flow_node);
+  }
+}
```

Comment:
What's the purpose of this? Are we adding it because it helps with optimization? Maybe we should do the same for regular JIT codepaths?

Reply:
We need to move transposes of weights into the ONNX `init_net` for perf reasons. But things inside control flow will not move to the init_net just the top-level nodes. ONNX Transpose is safe to speculate (it just reverses dimensions). Torch tranpose is not (it checks dim==2, i think). Moving it outside of an if statement can cause errors in a previously correct program. Almost _every_ operator in the graph is unsafe to speculate in this way, so this pass is very specific.

Reply:
(Just a general comment, not to this PR) I'm not sure why the transpose in torch would be unsafe, since you know that it's input won't change because of the blocks that you take it out of.